File: qmc_uniform.f90
subroutine uniform_montecarlo(a,nmax,energy)
  implicit none
  double precision, intent(in)  :: a
  integer*8       , intent(in)  :: nmax 
  double precision, intent(out) :: energy

  integer*8        :: istep
  double precision :: norm, r(3), w

  double precision, external :: e_loc, psi

  energy = 0.d0
  norm   = 0.d0

  do istep = 1,nmax

     call random_number(r)
     r(:) = -5.d0 + 10.d0*r(:)

     w = psi(a,r)
     w = w*w

     energy = energy + w * e_loc(a,r)
     norm   = norm   + w

  end do

  energy = energy / norm

end subroutine uniform_montecarlo

program qmc
  implicit none
  double precision, parameter :: a     = 0.9
  integer*8       , parameter :: nmax  = 100000
  integer         , parameter :: nruns = 30

  integer          :: irun
  double precision :: X(nruns)
  double precision :: ave, err

  do irun=1,nruns
     call uniform_montecarlo(a, nmax, X(irun))
  enddo

  call ave_error(X, nruns, ave, err)

  print *, 'E = ', ave, '+/-', err
end program qmc

File: qmc_uniform.c
#include "hydrogen.h"
#include "qmc_stats.h"
#include <stdlib.h>
#include <time.h>

double uniform_montecarlo(double a, const int nmax) {
      double r[3], energy, norm, w, rnd;

      norm = 0.0;
      energy = 0.0;
      for (int i = 0; i < nmax; ++i) {
	    for (int j = 0; j < 3; ++j) {
		  rnd = (double) rand() / RAND_MAX;
		  r[j] = -5.0 + 10.0 * rnd;
	    }
	    w = psi(a, r, 3) * psi(a, r, 3);
	    norm += w;
	    energy += w * e_loc(a, r, 3);
      }

      return energy / norm;

}

int main() {
      const double a = 1.2;
      const long nmax = 1e5;
      int nruns = 30;

      srand(time(NULL));

      double x[nruns], obs[2];

      for (int i = 0; i < nruns; ++i) {
	    x[i] = uniform_montecarlo(a, nmax);
      }

      ave_error(x, nruns, obs);

      printf("E = %.5lf +/- %.5lf\n", obs[0], obs[1]);
      
      return 0;
}

